### PR TITLE
Dont reset SharedRealm ptr

### LINKED ISSUE #4471: Don't reset the `SharedRealm.nativePtr` in `close()`
#### Goal

> What do you want to achieve?
I want to get List<Group> from realm db.
#### Expected Results

> 
Get filled list.

#### Actual Results
```
04-11 15:48:04.274 7158-7158/org.kaaproject.kaa.examples.gpiocontrol A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 7158 (les.gpiocontrol)
04-11 15:48:04.376 192-192/? A/DEBUG: pid: 7158, tid: 7158, name: les.gpiocontrol  >>> org.kaaproject.kaa.examples.gpiocontrol <<<
04-11 15:48:04.438 192-192/? A/DEBUG:     #00 pc 0002ace6  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/lib/arm/librealm-jni.so (Java_io_realm_internal_SharedRealm_nativeIsInTransaction+21)
04-11 15:48:04.439 192-192/? A/DEBUG:     #01 pc 011b6265  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (boolean io.realm.internal.SharedRealm.nativeIsInTransaction(long)+88)
04-11 15:48:04.439 192-192/? A/DEBUG:     #02 pc 011b7b91  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (boolean io.realm.internal.SharedRealm.isInTransaction()+68)
04-11 15:48:04.439 192-192/? A/DEBUG:     #03 pc 00d8fe15  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void io.realm.internal.Collection$Iterator.<init>(io.realm.internal.Collection)+296)
04-11 15:48:04.439 192-192/? A/DEBUG:     #04 pc 0119fe53  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void io.realm.OrderedRealmCollectionImpl$RealmCollectionIterator.<init>(io.realm.OrderedRealmCollectionImpl)+102)
04-11 15:48:04.439 192-192/? A/DEBUG:     #05 pc 014ce709  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (java.util.Iterator io.realm.OrderedRealmCollectionImpl.iterator()+76)
04-11 15:48:04.439 192-192/? A/DEBUG:     #06 pc 01603e99  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (java.util.Iterator io.realm.RealmResults.iterator()+60)
04-11 15:48:04.439 192-192/? A/DEBUG:     #07 pc 00f23f41  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (java.util.List org.kaaproject.kaa.examples.gpiocontrol.utils.Utils.getViewDeviceListFromGroup(java.util.List)+100)
04-11 15:48:04.439 192-192/? A/DEBUG:     #08 pc 00f23805  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (java.util.List org.kaaproject.kaa.examples.gpiocontrol.utils.Utils.getMockedHeaderList(org.kaaproject.kaa.examples.gpiocontrol.storage.Repository)+128)
04-11 15:48:04.439 192-192/? A/DEBUG:     #09 pc 01580557  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void org.kaaproject.kaa.examples.gpiocontrol.screen.deviceSwitchManagement.DeviceSwitchManagementFragment.setupRecyclerView(android.content.Context, android.os.Bundle)+442)
04-11 15:48:04.439 192-192/? A/DEBUG:     #10 pc 015823fd  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (android.view.View org.kaaproject.kaa.examples.gpiocontrol.screen.deviceSwitchManagement.DeviceSwitchManagementFragment.onCreateView(android.view.LayoutInflater, android.view.ViewGroup, android.os.Bundle)+568)
04-11 15:48:04.439 192-192/? A/DEBUG:     #11 pc 00a4d889  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (android.view.View android.support.v4.app.Fragment.performCreateView(android.view.LayoutInflater, android.view.ViewGroup, android.os.Bundle)+132)
04-11 15:48:04.439 192-192/? A/DEBUG:     #12 pc 00f9b179  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.moveToState(android.support.v4.app.Fragment, int, int, int, boolean)+6196)
04-11 15:48:04.439 192-192/? A/DEBUG:     #13 pc 00f98f6d  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(android.support.v4.app.Fragment)+352)
04-11 15:48:04.439 192-192/? A/DEBUG:     #14 pc 00f9956d  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.moveToState(int, boolean)+456)
04-11 15:48:04.440 192-192/? A/DEBUG:     #15 pc 00f841ff  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.BackStackRecord.executeOps()+1378)
04-11 15:48:04.440 192-192/? A/DEBUG:     #16 pc 00f8c44b  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.executeOps(java.util.ArrayList, java.util.ArrayList, int, int)+462)
04-11 15:48:04.440 192-192/? A/DEBUG:     #17 pc 00f8c945  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.executeOpsTogether(java.util.ArrayList, java.util.ArrayList, int, int)+1128)
04-11 15:48:04.440 192-192/? A/DEBUG:     #18 pc 00f8e4fd  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(java.util.ArrayList, java.util.ArrayList)+944)
04-11 15:48:04.440 192-192/? A/DEBUG:     #19 pc 00f96c17  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (boolean android.support.v4.app.FragmentManagerImpl.execPendingActions()+218)
04-11 15:48:04.440 192-192/? A/DEBUG:     #20 pc 00a52cb1  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (boolean android.support.v4.app.FragmentController.execPendingActions()+76)
04-11 15:48:04.440 192-192/? A/DEBUG:     #21 pc 016dcae5  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v4.app.FragmentActivity.onStart()+296)
04-11 15:48:04.440 192-192/? A/DEBUG:     #22 pc 01748fad  /data/app/org.kaaproject.kaa.examples.gpiocontrol-2/oat/arm/base.odex (offset 0x9e7000) (void android.support.v7.app.AppCompatActivity.onStart()+56)
```
>  E.g. full stack trace with exception

#### Steps & Code to Reproduce
Here is the full project https://github.com/iamtodor/gpio-control/tree/realm-report
It appearances once `DeviceSwitchManagementFragment` will start
> Describe your current debugging efforts.
If I use 
`List<Group> groupList = instance.where(Group.class).findAll();` - app crashes with the stacktrace, that I attached above
either I use this option `List<Group> groupList = instance.copyFromRealm(instance.where(Group.class).findAll());` getting data is okay, but after all operations I have to write within transaction `realm.insertOrUpdate(group);`

#### Code Sample

```java
https://github.com/iamtodor/gpio-control/tree/realm-report

> Your code here. Bigger samples should ideally be as separate Android Studio project, 
> in gists/repositories or privately at help@realm.io)

```

#### Version of Realm and tooling
Realm version(s): 3.0.0

Realm sync feature enabled: no

Android Studio version: 2.2.3

Which Android version and device: Nexus 7 2013, 6.0.1

### PR DESCRIPTION
**Pull request by beeender**
- Keep the SharedRealm ptr valid when close it. To allow the is_closed
  checking could throw in the Object Store and it can be converted to a
  friendly Java exception.
- Check if SharedRealm is closed when create collection iterator. Throw
  an ISE if it is. Fix #4471 .

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java [hunk 1]
```diff
@@ -368,12 +368,9 @@ public void close() {
...
         synchronized (context) {
-            if (nativePtr != 0) {
-                nativeCloseSharedRealm(nativePtr);
-                // It is OK to clear the nativePtr. It has been saved to the NativeObjectReference when adding to the
-                // context.
-                nativePtr = 0;
-            }
+            nativeCloseSharedRealm(nativePtr);
+            // Don't reset the nativePtr since we still rely on Object Store to check if the given SharedRealm ptr
+            // is closed or not.
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: `SharedRealm.isClosed()` should be updated as well.
No need to check `nativePtr` is null.

```
     public boolean isClosed() {
         return nativePtr == 0 || nativeIsClosed(nativePtr);
     }
 ```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 ### Enhancements
 
+* A ISE will be thrown when starting iterating `OrderedRealmCollection` if the Realm is closed (#4471).
```
Comment (zaki50) [inline CHANGELOG.md]: A -> An

Comment (dalinaum) [inline CHANGELOG.md]: An ISE

Comment (Zhuinden) [inline CHANGELOG.md]: Please just say `IllegalStateException` this is completely not obvious.

Comment (zaki50) [inline CHANGELOG.md]: I think it's a bugfix.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 ### Enhancements
 
-* An ISE will be thrown when starting iterating `OrderedRealmCollection` if the Realm is closed (#4471).
+* An `IllegalStateException` will be thrown when starting iterating `OrderedRealmCollection` if the Realm is closed (#4471).
```
Comment (Zhuinden) [inline CHANGELOG.md]: Thank you! 😄

Comment (dalinaum) [inline CHANGELOG.md]: @Zhuinden Thanks for your review.

**Pull request merged**